src/odemis/driver/keysight.py

```diff
@@ -0,0 +1,333 @@
+            while "\n" != ch:
+                if "\r" != ch:
+                    resp.append(ch)
+                ch = self._tcp_sock.recv(1).decode()
+            return ''.join(resp)
+
+    def sendCmd(self, cmd_str, wait_for_complete=True):
+        # Send the given command to the instrument.
+        if self._tcp_sock is not None:  # pass if not connected
+            cmd_str = cmd_str.rstrip()
+            cmd_str += '\n'
+            cmd_byte = cmd_str.encode()
+            self._tcp_sock.sendall(cmd_byte)  # send command
+
+
+class Keysight33622ASimulator(object):
+    """
+    Simulates a Keysight 33622A waveform generator
+    """
+
+    def __init__(self, timeout=1):
+        self.timeout = timeout
+        self._output_buf = ""  # what the commands sends back to the "host computer"
+        self._input_buf = ""  # what we receive from the "host computer"
+
+        self._output_states = {
+            'channel1': {
+                'tracking': 'on',  # on, off, invert
+                'voltage_low': -5.0,  # V
+                'voltage_high': 5.0,  # V
+                'voltage_limit_low': -5.0,  # V
+                'voltage_limit_high': 5.0,  # V
+                'period': 150,  # 1/frequency in Hz (1 μHz to 50 MHz)
+                'duty_cycle': 50  # percentage
+            },
+            'channel2': {
+                'tracking': 'on',  # on, off, invert
+                'voltage_low': -5.0,  # V
+                'voltage_high': 5.0,  # V
+                'voltage_offset': 0.0,  # V
+                'voltage_limit_low': -5.0,  # V
+                'voltage_limit_high': 5.0,  # V
+                'period': 150,  # 1/frequency in Hz (1 μHz to 50 MHz)
+                'duty_cycle': 75  # percentage
+            },
+        }
+
+    def write(self, data):
+        self._input_buf += data
+        msgs = self._input_buf.split("\n")
+        for m in msgs[:-1]:
+            self._parseMessage(m)  # will update _output_buf
+
+        self._input_buf = msgs[-1]
+
+    def read(self, size=1):
+        ret = self._output_buf[:size]
+        self._output_buf = self._output_buf[len(ret):]
+
+        if len(ret) < size:
+            # simulate timeout
+            time.sleep(self.timeout)
+        return ret
+
+    def close(self):
+        # using read or write will fail after that
+        del self._output_buf
+        del self._input_buf
+
+    def _sendAnswer(self, ans):
+        self._output_buf += "%s\n" % (ans,)
+
+    def _parseMessage(self, msg):
+        """
+        msg (str): the message to parse
+        return None: self._output_buf is updated if necessary
+        """
+        logging.debug("SIM: parsing %s", to_str_escape(msg))
+        msg = msg.decode("latin1").strip()  # remove leading and trailing whitespace
+        msg = "".join(msg.split())  # remove all space characters
+
+        if msg == "*IDN?":
+            self._sendAnswer("Agilent Technologies,33622A,MY59003187,A.02.03-3.15-03-64-0")
```

Comment:
Please return a string which clearly indicate it's a simulator. For instance, the serial number can be "FAKE".

Reply:
I wanted to make it as close to the real hardware as possible so I've placed the exact response from the hardware. 

But it makes sense to indicate that the response is coming from the simulator. Will update the string then!